lib/SimpleSAML/XHTML/IdPDisco.php

```diff
@@ -600,13 +600,30 @@ public function handleRequest()
         $language = $translator->getLanguage()->getLanguage();
         $tryLanguages = [0 => $language, 1 => $defaultLanguage, 2 => $fallbackLanguage];
 
+        $t->data['idplist'] = $this->getPreferredTranslations($idpList, $tryLanguages);
+        $t->data['preferredidp'] = $preferredIdP;
+        $t->data['return'] = $this->returnURL;
+        $t->data['returnIDParam'] = $this->returnIdParam;
+        $t->data['entityID'] = $this->spEntityId;
+        $t->data['urlpattern'] = htmlspecialchars(Utils\HTTP::getSelfURLNoQuery());
+        $t->data['rememberenabled'] = $this->config->getBoolean('idpdisco.enableremember', false);
+        $t->show();
+    }
+
+
+    /**
+     * @param array $idpList
+     * @return array
+     */
+    private function getPreferredTranslations(array $idpList, array $tryLanguages): array
```

Comment:
I think this is quite misleading. What this method is doing is returning a list of IdPs, sorted by name, where their names are translated based on the preferred translation for the given request. I don't think _getPreferrredTranslations()_ correctly describes that, but instead it sounds like it's just returning the preferred translations for some given IdPs.

I would actually suggest refactoring this code. There's a lot of repetition in there, and testing is quite hard, so we could benefit from routing to feed requests and get responses out of it.

I'm going to merge @m0ark's commit into 1.18, and we can keep working on this meanwhile.